### DISCUSSION TITLE
PR description is not helpful in understanding why a PR is "immortal"

### DISCUSSION THREAD

**Discussion by ecraig12345** [renovatebot/renovate#15628] [Suggest an Idea]
I've been trying to understand why a Renovate PR updating a custom group with React plus related types is being marked as "immortal." There are a couple things which I think could help make this less confusing.

The PR description's "immortal" bullet point has a link "get config help" which links to discussions here, but it would be helpful if it could additionally link to relevant docs to help people understand what's going on.

What I didn't realize until searching Renovate's code for the term "immortal" (I was getting pretty desperate) is that this is handled in code as the config option `recreateClosed`, and the [docs for that option](https://docs.renovatebot.com/configuration-options/#recreateclosed) mention a couple special cases which (partly) helped me understand what's happening. Problem is, someone couldn't find this by searching within the *docs* for the term "immortal" because that term isn't used in the docs.

So I think at minimum it would be helpful to add a link to the existing `recreateClosed` docs from the "immortal" bullet point in the PR description. Or even better, extract that information onto a separate page with more detail about what could cause a PR to be "immortal" and how to fix it, and link to that (also reduces repetitive discussion questions).

(I could definitely help with adding the word "immortal" to the existing `recreateClosed` docs and linking from the PR bullet point if that's desired, but I probably don't understand the issue or the doc structure well enough to make a separate page.)

**Comment by rarkins** [comment]
PRs welcome: https://github.com/renovatebot/renovate/tree/main/docs/usage

**Reply by HonkingGoose** [reply]
I'm opening an issue first, so we can discuss what we need, before I start work. 😉 See:

- #15797

**Comment by dzearing** [comment]
I'm really curious on the original motivation of "immortal". When would someone want a PR to be constantly re-opened when they close it? And why is React defaulted to that? It seems highly penalizing when you have no bleeding need to upgrade React as it releases a new, potentially destabilizing, release. This is especially apparent when the confidence metrics are low.

We have been thinking about this problem internally. I think groups of dependencies have an expiration date; computed by the delta between time the first update was available to now and comparing it with a "max shelf life" date. E.g. If React releases a new update, a PR might be welcome, and we'd choose to either schedule the upgrade, or close it for another time. We'd expect it to go to the blocked list on the dashboard issue. But, after some time, it's no longer "ok" to put it off. This is where immortal could be useful. It shows up, and keeps re-opening to enforce a company wide policy on the max life time of an aged dependency.

Another scenario where I could see immortal being used is to diffuse a vulnerable dependency. It's a nice compliment to a big red security banner.

In both cases, the comments would need to be clear what the intention of the PR being reopened was. Wording might need to be different in each.

**Reply by rarkins** [reply]
It's not because we want to make them immortal for some philosophical reason ("don't ignore this, it's good for you!") but instead because we can't *ignore* them after they're closed.

Renovate uses branch name + PR title like a cache key. If the same key exists and was closed unmerged, then we don't reopen.

But take the case where someone has a "All non-major updates" PR. If they closed that, and then we ignore based on that PR title, it means they'll never get a non-major update again. Your react case would be similar.

So we only can do ignoring if there's a unique version (such as "to v16.1.2", or "to v16") in the title which is safe to ignore on.

So the problem comes when there's groups of updates which have differing versions. Then the update becomes "Update react (major)" - which is not safely ignorable - instead of "Update react to v16".

In the future we may embed metadata in each PR identifying the exact files and packages + versions that PR contains. Then we could allow such PRs to be closed/ignored but then as soon as there's any chance to files/packages/versions being updated then we'd be cached busted and create a new PR.

BTW in my own workflows I default all major updates to require dependency dashboard approval. That way I open them on demand, and if we need to close them then they don't get recreated automatically - they go back to requiring dashboard approval.

**Reply by rarkins** [reply]
If the above is helpful then please feel free to copy/paste/paraphrase it into our docs!
